controllers/volumereplicationgroup_controller.go

```diff
@@ -197,77 +207,284 @@ func (v *VolumeReplicationGroupReconciler) Reconcile(ctx context.Context, req ct
+		log.Errorf("failed to remove the finalizer from vrg: %s/%s", v.instance.Name, v.instance.Namespace)
+
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// stop reconciliation as the item is being deleted
+	return ctrl.Result{}, nil
+}
+
+// processAsPrimary reconciles the current instance of VRG as primary
+/* Primary:
+  - For each PVC
+	- If not bound, continue but requeue
+    * Handle PVC being Deleted?
+	- Check `vr-replicated` finalizer, if found skip to checking status of PVC
+	- Add `vr-protection` finalizer to PVC (to delete/update VR when PVC is deleted)
+    - [Find|Create] VR for PVC as Primary
+	- Backup PV bound to PVC
+	- Add `vr-replicated` finalizer to PVC
+  - For each PVC
+    - Check if VR status has reached Primary state
+		- If not, update status as progressing
+  - Report status as completed, if not progressing for any PVC
+* For PVCs being deleted with VRG as primary, we need to handle garbage collection of PVs
+*/
+func (v *VRGInstance) processAsPrimary() (ctrl.Result, error) {
+	if err := v.addFinalizer(vrgFinalizerName); err != nil {
+		log.Errorf("failed to add the finalizer to vrg: %s/%s", v.instance.Name, v.instance.Namespace)
+
+		return ctrl.Result{Requeue: true}, nil
+	}
+
 	requeue := false
+	requeue = v.handlePersistentVolumeClaims()
 
-	for _, pvc := range pvcList.Items {
-		volRep := &volrep.VolumeReplication{}
+	if requeue {
+		return ctrl.Result{Requeue: requeue}, nil
+	}
 
-		err = v.Get(ctx, types.NamespacedName{Name: pvc.Name, Namespace: pvc.Namespace}, volRep)
-		if err != nil {
-			if errors.IsNotFound(err) {
-				if err = v.createVolumeReplicationCRForPVC(ctx, pvc); err == nil {
-					continue
-				}
-			}
+	return ctrl.Result{}, nil
+}
 
-			// requeue on failure to ensure any PVC not having a corresponding VR CR
-			requeue = true
+// processAsSecondary reconciles the current instance of VRG as secondary
+/* Secondary:
+  - For each PVC
+	- If not bound, continue
+	- Ensure `vr-protection` finalizer is present for PVC, if not continue
+	- Ensure PVC not in use
+		https://kubernetes.io/docs/concepts/storage/persistent-volumes/#storage-object-in-use-protection
+		- Continue to next PVC, but requeue VRG
+	- Ensure PVC being deleted; to prevent inadvertent use by pods as we change state to Secondary
+		- Ideally, the Subscription and hence the PVC would be deleted, as we proceed to update VRG
+		as secondary and then delete VRG
+		- Continue to next PVC, but requeue VRG
+	- Find VR for PVC and update to Secondary
+		- NOTE: There is a potential gap here, if VR did not get created in time for this PVC
+		- Check if `vr-replicated` is missing, of so create VR, and backup PV etc.
+  - For each PVC
+    - Check if VR status has reached Secondary state
+		- If not, update status as progressing
+  - Report status as completed, if not progressing for any PVC
+*/
+func (v *VRGInstance) processAsSecondary() (ctrl.Result, error) {
+	return v.processAsPrimary()
```

Comment:
(nit) This is a minor comment. Can be done in a subsequent separate PR as well.

Since, the handling of a secondary VRG has to be done by adding more logic (some of the logic mentioned in the above algorithm is available already), do we want to simply return success instead of doing "return v.processAsPrimary()" ?

Reply:
The 2 comments (which on hindsight I should not have added in this commit) are thought in progress.

With that said, I wanted to retain current behavior with this commit, which is to use the VRG state to update the VR state, which flows through the primary handling, hence just calling handle as primary here.